### PR TITLE
add cookiecutter generation for InvenioILS

### LINKED ISSUE #1167: Create a cookiecutter generation for InvenioILS
1. Read about [cookiecutter](https://github.com/cookiecutter/cookiecutter)
2. Prepare a cookiecutter template for InvenioILS as done for [InvenioRDM](https://github.com/inveniosoftware/cookiecutter-invenio-rdm)
3. The goal is to be able to:
  a. `cookiecutter` a new instance
  b. run `/scripts/setup && /scripts/server` and `npm install && npm start`
  
4. Main parts to be added:
  a. Invenio config (invenio.cfg as done in InvenioRDM)
  b. the UI config.js and CSS styles
  c. vocabularies JSON files (an empty folder where to put it)
  d. email template HTML files (an empty folder where to put it)

### PR DESCRIPTION
**Pull request by anikachurilova**
* add possibility to cookiecutter a new InvenioILS instance
* add possibility to build docker images for services and for the whole application
* add script to test the cookiecutter
* closes https://github.com/inveniosoftware/invenio-app-ils/issues/1167

### CODE REVIEW

File: AUTHORS.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    Copyright (C) 2022 CERN.
+
+    Invenio is free software; you can redistribute it and/or modify it
+    under the terms of the MIT License; see LICENSE file for more details.
+
+
+Contributors
+============
+
+- Anna Churilova
```
Comment (ntarocco) [inline AUTHORS.rst:13]: 🚀

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+==============================
+ Cookiecutter-Invenio-ILS
+==============================
+
+.. image:: https://img.shields.io/github/license/inveniosoftware/cookiecutter-invenio-ils.svg
+        :target: https://github.com/inveniosoftware/cookiecutter-invenio-ils/blob/master/LICENSE
+
+.. image:: https://img.shields.io/travis/inveniosoftware/cookiecutter-invenio-ils.svg
+        :target: https://travis-ci.org/inveniosoftware/cookiecutter-invenio-ils
```
Comment (ntarocco) [inline README.rst]: We don't use `Travis` anymore (it is apparently wrong in RDM too!), but GitHub CI instead: https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/adding-a-workflow-status-badge
Can you please fix RDM too?

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+Quickstart
+----------
+
+Install the latest Cookiecutter if you haven't installed it yet::
+
+    pip install -U cookiecutter
+
+Generate your Invenio ILS instance::
+
+    cookiecutter https://github.com/inveniosoftware/cookiecutter-invenio-ils.git -c v1.0
```
Comment (ntarocco) [inline README.rst]:
```suggestion
    cookiecutter https://github.com/inveniosoftware/cookiecutter-invenio-ils.git -c v1.0.0rc1
```

File: README.rst [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+More detailed steps can be found in the https://inveniosoftware.org/#getstarted
```
Comment (ntarocco) [inline README.rst]: The link is not correct, shall we put <https://invenioils.docs.cern.ch> instead?

File: cookiecutter.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    "project_shortname": "{{ cookiecutter.project_name | slugify() }}",
+    "project_site": "{{ cookiecutter.project_shortname }}.com",
+    "github_repo": "{{ cookiecutter.project_shortname }}/{{ cookiecutter.project_shortname }}",
+    "description": "{{ cookiecutter.project_name }} InvenioILS Instance",
+    "author_name": "CERN",
+    "author_email": "info@{{ cookiecutter.project_site }}",
+    "year": "{% now 'local', '%Y' %}",
+    "python_version": ["3.7", "3.8", "3.9"],
+    "database": ["postgresql"],
+    "search": ["elasticsearch7", "opensearch1"],
```
Comment (ntarocco) [inline cookiecutter.json]: For now, we did not upgrade to OpenSearch yet :(
```suggestion
    "search": ["elasticsearch7"],
```

File: run-tests.sh [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+trap finish EXIT
+
+PROJECT_NAME="my-site"
+export PROJECT_NAME
+
+cookiecutter --no-input -o "$WORKDIR" . \
+    project_name=${PROJECT_NAME} \
+    database=${COOKIECUTTER_DATABASE:-postgresql} \
+    elasticsearch=${COOKIECUTTER_ELASTICSEARCH:-elasticsearch6} \
```
Comment (ntarocco) [inline run-tests.sh]:
```suggestion
    elasticsearch=${COOKIECUTTER_ELASTICSEARCH:-elasticsearch7} \
```

File: scripts/bootstrap [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+cd ${PROJECTDIR}
+pipfile_lock_path="./Pipfile.lock"
+
+if [ ! -f $pipfile_lock_path ]; then
+    echo "'Pipfile.lock' not found. Generating via 'pipenv lock --dev'..."
+    pipenv lock --dev --pre 
+fi
+
+# Installs all packages specified in Pipfile.lock
+pipenv sync --dev --pre
```
Comment (ntarocco) [inline scripts/bootstrap]:
```suggestion
pipenv sync
```

File: scripts/wait-for-services.sh [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+_db_check(){ docker-compose exec --user postgres db bash -c "pg_isready" &>/dev/null; }
+check_ready "Postgres" _db_check
+
+_search_check(){ curl --output /dev/null --silent --head --fail http://localhost:9200 &>/dev/null; }
+if [[ ${COOKIECUTTER_SEARCH} == "elasticsearch7" ]]
+then
+    check_ready "Elasticsearch" _search_check
+else
+    check_ready "OpenSearch" _search_check
```
Comment (ntarocco) [inline scripts/wait-for-services.sh]: please remove the `OpenSearch` :)

File: {{cookiecutter.project_shortname}}/Pipfile [hunk 9]
```diff
@@ -0,0 +1,20 @@
+[[source]]
+name = "pypi"
+url = "https://pypi.org/simple"
+verify_ssl = true
+
+[dev-packages]
+check-manifest = ">=0.25"
+
+[packages]
+invenio-app-ils = {extras = ["postgresql", "elasticsearch7", "lorem"], version = "~=1.0.0a67"}
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/Pipfile]: we will release `rc1`
```suggestion
invenio-app-ils = {extras = ["postgresql", "elasticsearch7", "lorem"], version = "~=1.0.0rc1"}
```

File: {{cookiecutter.project_shortname}}/README.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
+# {{ cookiecutter.project_name }}
+
+Welcome to your InvenioILS instance.
+
+## Getting started
+
+Run the following commands in order to start your new InvenioILS instance:
+
+**TODO**
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/README.md]: TODO?

File: {{cookiecutter.project_shortname}}/README.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+## Overview
+
+Following is an overview of the generated files and folders:
+
+| Name                      | Description                                                                                                      |
+| ------------------------- | ---------------------------------------------------------------------------------------------------------------- |
+| `docker`                  | Example configuration for NGINX and uWSGI. Consists Dockerfiles for building backend and frontend docker images. |
+| `templates`               | Folder for your Jinja templates.                                                                                 |
+| `translations`            | Folder for page translations.                                                                                    |
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/README.md]: I would remove this for the moment
```suggestion

```

File: {{cookiecutter.project_shortname}}/README.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+| `vocabularies`            | Folder for vocabularies (mappings, schemas, etc).                                                                |
+| `docker-compose.full.yml` | Example of a full infrastructure stack.                                                                          |
+| `docker-compose.yml`      | Backend services needed for local development.                                                                   |
+| `docker-services.yml`     | Common services for the Docker Compose files.                                                                    |
+| `invenio.cfg`             | The Invenio application configuration.                                                                           |
+| `Pipfile`                 | Python requirements installed via [pipenv](https://pipenv.pypa.io)                                               |
+| `Pipfile.lock`            | Locked requirements (generated on first install).                                                                |
+| **TODO**                  | **TODO**                                                                                                         |
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/README.md]: TODO

File: {{cookiecutter.project_shortname}}/invenio.cfg [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+###############################################################################
+# Data model extensions
+###############################################################################
+ILS_RECORDS_METADATA_EXTENSIONS = {}
+
+###############################################################################
+# Vocabularies
+###############################################################################
+#ILS_VOCABULARIES = []
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/invenio.cfg:130]: question: is this correct?

File: {{cookiecutter.project_shortname}}/templates/invenio_app_ils/notifications/{{cookiecutter.project_shortname}}_footer.html [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+{% block footer_html %}
+<br/><br/>
+Check your library loans and requests in your <a href="{{ spa_routes.HOST }}{{ spa_routes.PATHS["profile"] }}">profile page</a>.
+<br/><br/>
+Kind regards, <br/><br/>
+
+CERN Library <br/>
+Location: building 52/1-052 on the Meyrin campus <br/>
+<a href="https://catalogue.library.cern/opening-hours"> Opening hours</a> <br/>
+Email: library.desk@cern.ch <br/>
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/templates/invenio_app_ils/notifications/{{cookiecutter.project_shortname}}_footer.html]: Please replace all `library.desk@cern.ch` with `library.desk@{{ cookiecutter.project_shortname }}` (to be tested)

File: {{cookiecutter.project_shortname}}/templates/invenio_app_ils/notifications/{{cookiecutter.project_shortname}}_footer.html [hunk 15]
```diff
@@ -0,0 +1,25 @@
+{% raw %}
+{% block footer_plain %}
+Check your library loans and requests in your profile page <{{ spa_routes.HOST }}{{ spa_routes.PATHS["profile"] }}>.
+
+Kind regards,
+
+CERN Library
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/templates/invenio_app_ils/notifications/{{cookiecutter.project_shortname}}_footer.html]: Let's remove this, and put {{ cookiecutter.project_shortname }} only, nothing below.

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/elements/label.overrides [hunk 16]
```diff
@@ -0,0 +1,19 @@
+/*******************************
+         Site Overrides
+*******************************/
+
+/* Frontsite */
+@{fs-parent-selector} {
+    .standard-number {
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/elements/label.overrides]: I think we can leave this part empty, as for the others

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/elements/container.overrides [hunk 17]
```diff
@@ -0,0 +1,12 @@
+/*******************************
+         Site Overrides
+*******************************/
+
+/* Frontsite */
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/elements/container.overrides]: I think we can leave this part empty, as for the others

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/globals/site.overrides [hunk 18]
```diff
@@ -0,0 +1,61 @@
+/*******************************
+   Site Overrides - CDS ILS
+*******************************/
+
+h2.logo,h3.logo {
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/globals/site.overrides]: I think we can leave this part empty, as for the others

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/globals/site.overrides [hunk 19]
```diff
@@ -0,0 +1,61 @@
+/*******************************
+   Site Overrides - CDS ILS
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/globals/site.overrides]:
```suggestion
   Site Overrides
```

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/globals/site.variables [hunk 20]
```diff
@@ -0,0 +1,3 @@
+/*******************************
+   Site.Variables - CDS ILS
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/globals/site.variables]: Please remove any trace of `CDS`

```suggestion
   Site.Variables
```

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/modules/accordion.overrides [hunk 21]
```diff
@@ -0,0 +1,16 @@
+/*******************************
+        User Overrides
+*******************************/
+
+.backoffice{
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/modules/accordion.overrides]: I think we can leave this part empty, as for the others

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/modules/search.overrides [hunk 22]
```diff
@@ -0,0 +1,7 @@
+/*******************************
+         Site Overrides
+*******************************/
+.ui.large.search.clean-search{
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/modules/search.overrides]: I think we can leave this part empty, as for the others

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/views/card.overrides [hunk 23]
```diff
@@ -0,0 +1,15 @@
+/*******************************
+    User Variable Overrides
+*******************************/
+
+/* Frontsite */
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/views/card.overrides]: I think we can leave this part empty, as for the others

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/views/statistic.overrides [hunk 24]
```diff
@@ -0,0 +1,17 @@
+/*******************************
+    User Variable Overrides
+*******************************/
+
+.ui.statistic.import-statistic{
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/site/views/statistic.overrides]: I think we can leave this part empty, as for the others

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/theme.config [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+╚══██╔══╝██║  ██║██╔════╝████╗ ████║██╔════╝██╔════╝
+   ██║   ███████║█████╗  ██╔████╔██║█████╗  ███████╗
+   ██║   ██╔══██║██╔══╝  ██║╚██╔╝██║██╔══╝  ╚════██║
+   ██║   ██║  ██║███████╗██║ ╚═╝ ██║███████╗███████║
+   ╚═╝   ╚═╝  ╚═╝╚══════╝╚═╝     ╚═╝╚══════╝╚══════╝
+
+*/
+
+/*******************************
+        Theme Selection - CDS ILS
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/theme.config]:
```suggestion
        Theme Selection - {{cookiecutter.project_shortname}}
```

File: {{cookiecutter.project_shortname}}/ui/src/semantic-ui/theme.less [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+
+/*******************************
+             Mix-ins
+*******************************/
+
+/*------------------
+       Fonts
+-------------------*/
+
+.loadFonts() when (@importGoogleFonts) {
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/ui/src/semantic-ui/theme.less:54]: Can you please remove this part? I thought we have removed it...

File: {{cookiecutter.project_shortname}}/invenio.cfg (hunk 13 above)
Comment (anikachurilova) [reply {{cookiecutter.project_shortname}}/invenio.cfg:130]: Yes, it's the name of the variable where user will put his list of available vocabularies

File: {{cookiecutter.project_shortname}}/README.md [hunk 27]
```diff
@@ -0,0 +1,62 @@
+# {{ cookiecutter.project_name }}
+
+Welcome to your InvenioILS instance.
+
+## Getting started
+
+To start with, create a new Python `virtualenv`:
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/README.md:7]: I think that with `pipenv` we don't need the virtualenv

File: {{cookiecutter.project_shortname}}/setup.sh [hunk 28]
```diff
@@ -0,0 +1,16 @@
...
+# This file is part of Invenio.
+#
+# Copyright (C) 2022 CERN.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+docker-compose up -d
+pipenv install
+invenio setup --verbose
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/setup.sh:14]: prefix with `pipenv run`?

File: {{cookiecutter.project_shortname}}/README.md [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+```console
+mkvirtualenv my-site
+```
+
+Run the `./setup.sh` script in order to start your new InvenioILS instance.
+
+Run the backend:
+
+```console
+FLASK_ENV=development invenio run --cert docker/backend/test.crt --key docker/backend/test.key
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/README.md:18]: prefix with `pipenv run`?

File: {{cookiecutter.project_shortname}}/docker-services.yml [hunk 30]
```diff
@@ -0,0 +1,80 @@
...
+      - "POSTGRES_DB={{cookiecutter.project_shortname}}"
+    ports:
+      - "5432:5432"
+  mq:
+    image: rabbitmq:3-management
+    ports:
+      - "15672:15672"
+      - "5672:5672"
+  es:
+    # to use ES6: docker.elastic.co/elasticsearch/elasticsearch:6.8.8
```
Comment (ntarocco) [inline {{cookiecutter.project_shortname}}/docker-services.yml:52]: we can remove ES6

File: README.rst [hunk 31]
```diff
@@ -0,0 +1,54 @@
...
+    under the terms of the MIT License; see LICENSE file for more details.
+
+==============================
+ Cookiecutter-Invenio-ILS
+==============================
+
+.. image:: https://img.shields.io/github/license/inveniosoftware/cookiecutter-invenio-ils.svg
+        :target: https://github.com/inveniosoftware/cookiecutter-invenio-ils/blob/master/LICENSE
+
+.. image:: https://github.com/inveniosoftware/cookiecutter-invenio-ils/workflows/CI/badge.svg
```
Comment (ntarocco) [inline README.rst:15]: remove me

**Review by ntarocco — APPROVED** [review]

**Pull request merged**